contracts/zero-ex/contracts/src/features/MultiplexFeature.sol

```diff
@@ -0,0 +1,799 @@
+    ///      operations.
+    /// @param fillData Encodes the path of tokens, the sell amount,
+    ///        and the nested operations for this multi-hop fill.
+    /// @param minBuyAmount The minimum amount of the output token
+    ///        to buy. Reverts if this amount is not met.
+    /// @return outputTokenAmount The amount of the output token bought.
+    function multiHopFill(
+        MultiHopFillData memory fillData,
+        uint256 minBuyAmount
+    )
+        public
+        payable
+        override
+        returns (uint256 outputTokenAmount)
+    {
+        require(
+            fillData.tokens.length == fillData.calls.length + 1,
+            "MultiplexFeature::multiHopFill/MISMATCHED_ARRAY_LENGTHS"
+        );
+        IERC20TokenV06 outputToken = IERC20TokenV06(fillData.tokens[fillData.tokens.length - 1]);
+        uint256 senderBalanceBefore = outputToken.getTokenBalanceOf(msg.sender);
+        uint256 ethBalanceBefore = address(this).balance.safeSub(msg.value);
+
+        _multiHopFill(fillData, msg.value);
+
+        // RFC: I think these balance checks are arguably unnecessary,
+        // see reasoning above.
+
+        // The `outputTokenAmount` returned by `_multiHopFill` may not
+        // be fully accurate (e.g. due to some janky token and/or
+        // reentrancy situation).
+        outputTokenAmount = outputToken.getTokenBalanceOf(msg.sender)
+            .safeSub(senderBalanceBefore);
+        require(
+            outputTokenAmount >= minBuyAmount,
+            "MultiplexFeature::multiHopFill/UNDERBOUGHT"
+        );
+
+        uint256 ethBalanceAfter = address(this).balance;
+        require(
+            ethBalanceAfter >= ethBalanceBefore,
+            "MultiplexFeature::multiHopFill/OVERSPENT_ETH"
+        );
+        // Refund ETH
+        if (ethBalanceAfter > ethBalanceBefore) {
+            payable(msg.sender).transfer(ethBalanceAfter - ethBalanceBefore);
+        }
+    }
+
+    // Similar to FQT. If `fillData.sellAmount` is set to `type(uint256).max`,
+    // this is effectively a batch fill. Otherwise it can be set to perform a
+    // market sell of some amount. In the case of a market sell, the `sellAmount`
+    // values of the constituent wrapped calls can encode percentages of the
+    // full `fillData.sellAmount`. This is useful if e.g. the `_batchFill` call
+    // the second leg of a multi-hop fill, where the sell amount is not known
+    // ahead of time.
+    function _batchFill(BatchFillData memory fillData, uint256 totalEth)
+        public
+        returns (uint256 outputTokenAmount, uint256 remainingEth)
+    {
+        // Track the remaining ETH allocated to this call.
+        remainingEth = totalEth;
+        // Track the amount of input token sold.
+        uint256 soldAmount;
+        for (uint256 i = 0; i != fillData.calls.length; i++) {
+            // Check if we've hit our target.
+            if (soldAmount >= fillData.sellAmount) { break; }
```

Comment:
Is it possible we oversell? If the user signs off on selling 1 WETH then we should sell at most 1 ETH. Rounding should always favor the sender.

Reply:
I don't think it's possible to oversell (barring the taker having set an allowance on a malicious token contract or something), since the `_normalizeAmount` function clips the input token amount to the remaining sell amount. But we can revert if `soldAmount > fillData.sellAmount` I guess?

Reply:
Yeah, actually it could be good to return `sellAmount` and then assert `sellAmount <= fillData.sellAmount` in the entry functions, where we validate `boughtAmount >= minBoughtAmount`.

Reply:
Or, possibly even using a balance check. In that case, so long as the user is not transferring out intermediate tokens, audting the security would be very simple.

Reply:
idk, I'm still pretty convinced that even the output token balance check is unnecessary. If we look at all of the functions this feature can call, we should be able to trust the output token amount returned by all of them:
- `_sellToUniswap` we compute the output token amount. This is what we do in the UniswapFeature as well, where we do not perform any balance checks. ✅ 
- `_sellToLiquidityProvider` we use balance checks to compute the output token amount. ✅ 
- `_fillRfqOrder` has the same guarantees as filling an RFQ order through the NativeOrdersFeature ✅ 
- `_transformERC20` performs balance checks for the output token amount ✅ 
- WETH withdraw/deposit is obvious ✅ 
- By induction, we should be able to trust the output token amount of a nested `_multiHopFill` as well ✅

Reply:
Of course, if the user has set an allowance on some malicious token contract that gets called in this feature, all bets are off. But this is also the case for all our other features. So I don't think we'd be introducing any new security assumptions by not performing balance checks on the input token (or even the output token)

Reply:
@mzhu25 I agree with you but I think the optics and audits will go better with the check in place since very few people would be able to prove it in this manner. It also protects us if we violate these assumptions by updating a feature or w/e.

Reply:
I'm fine with leaving the output token balance check in, but I think checking balances for the input token or intermediate tokens would be overkill

Reply:
Could just have the input token check as part of a test to catch regressions.